object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java

```diff
@@ -554,6 +576,66 @@ private void mustCallTraverse(ControlFlowGraph cfg) {
     }
   }
 
+  private void checkAssignmentToOwningField(AssignmentNode assignmentNode) {
+    Node lhs = assignmentNode.getTarget();
+    Node rhs = assignmentNode.getExpression();
+    Element fieldElement = TreeUtils.elementFromTree(lhs.getTree());
+    Element enclosingElement = fieldElement.getEnclosingElement();
+    List<String> rhsMCAnno = getMustCallValue(TreeUtils.elementFromTree(rhs.getTree()));
+    AnnotationMirror rhsCalledMethodAnno = createCalledMethods(rhsMCAnno.toArray(new String[0]));
+    List<String> enclosingElAnno = getMustCallValue(enclosingElement);
+    boolean report = true;
+    if (enclosingElAnno != null
+        && (getMustCallValue(fieldElement).isEmpty()
+            || fieldElement.getModifiers().contains(Modifier.FINAL))) {
+      List<? extends Element> classElements = enclosingElement.getEnclosedElements();
+      for (Element element : classElements) {
+        if (element.getKind().equals(ElementKind.METHOD)) {
+          // TODO cause false positives if there is more than one value in the enclosingElAnno
+          if (element.getSimpleName().toString().equals(enclosingElAnno.get(0))) {
+            AnnotationMirror annotationMirror =
+                getDeclAnnotation(element, EnsuresCalledMethods.class);
+            if (annotationMirror == null) {
+              break;
+            }
+            Attribute values = ((Attribute.Compound) annotationMirror).getValue().values.get(0).snd;
```

Comment:
Please place some of this extraction logic into separate methods with appropriate names.  This is pretty hard to read.